native/src/main/scala/sqs4s/auth/Credentials.scala

```diff
@@ -0,0 +1,250 @@
+import cats.effect.implicits._
+import cats.effect.{Clock, Concurrent, Resource, Sync, Timer}
+import cats.implicits._
+import fs2._
+import org.http4s.Method.{GET, PUT}
+import org.http4s.circe.CirceEntityDecoder._
+import org.http4s.client.Client
+import org.http4s.client.dsl.Http4sClientDsl
+import org.http4s.implicits._
+import org.http4s.{Header, Response, Status}
+import sqs4s.auth.errors._
+
+import scala.concurrent.duration._
+
+trait Credentials[F[_]] {
+  def get: F[Credential]
+}
+
+sealed trait Credential {
+  def accessKey: String
+  def secretKey: String
+}
+
+case class BasicCredential(accessKey: String, secretKey: String)
+    extends Credential
+
+case class TemporarySecurityCredential(
+  accessKey: String,
+  secretKey: String,
+  sessionToken: String
+) extends Credential
+
+/**
+  * Implementation which follows AWS doc
+  * https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#credentials-default
+  */
+object Credentials {
+  def instance[F[_]](credF: F[Credential]): Credentials[F] =
+    new Credentials[F] {
+      def get: F[Credential] = credF
+    }
+
+  def basic[F[_]: Applicative](
+    accessKey: String,
+    secretKey: String
+  ): Credentials[F] =
+    Credentials.instance[F](
+      BasicCredential(accessKey, secretKey).pure[F].widen
+    )
+
+  /**
+    * Load Credentials in this order:
+    * Environment variables
+    * Java system properties
+    * Instance profile credentials– used on EC2 instances, and delivered through
+    * the Amazon EC2 metadata service.
+    *
+    * @param client http client
+    * @param ttl session token's time to live
+    * @param refreshBefore a small duration to refresh the token before it expires
+    * @tparam F an effect which represents the side effects
+    * @return
+    */
+  def chainResource[F[_]: Concurrent: Clock: Timer](
+    client: Client[F],
+    ttl: FiniteDuration = 6.hours,
+    refreshBefore: FiniteDuration = 5.minutes
+  ): Resource[F, Credentials[F]] = {
+    Resource {
+      envVar[F].map(cred => cred -> ().pure[F]).handleErrorWith(_ =>
+        sysProp[F].map(cred => cred -> ().pure[F])).handleErrorWith(_ =>
+        instanceMetadata[F](client, ttl, refreshBefore))
+    }
+  }
+
+  /**
+    * Create a resource of pure value Credential where the access key and secret
+    * key are static
+    *
+    * @param accessKey static access key
+    * @param secretKey static secret key
+    * @tparam F an effect which represents pure value
+    * @return a Resource of TemporarySecurityCredential
+    */
+  def basicResource[F[_]: Applicative](
+    accessKey: String,
+    secretKey: String
+  ): Resource[F, Credentials[F]] =
+    Resource.pure[F, Credentials[F]](instance(BasicCredential(
+      accessKey,
+      secretKey
+    ).pure[F].widen[Credential]))
+
+  def envVarResource[F[_]: Sync]: Resource[F, Credentials[F]] =
+    Resource.liftF(envVar[F])
+
+  private def envVar[F[_]: Sync]: F[Credentials[F]] =
```

Comment:
I believe this is one _(as well as the `sysProp` one)_ are wrong.
They will attempt to read the env variables or system properties in each use, whereas they should read them at initialization and then cache the results _(IMHO)_; this change is essential for my previous suggestion.